### PR TITLE
Streams

### PR DESCRIPTION
**Pull request by bserdar**
Stream result sets instead of passing them around in lists. This is the first step for dealing with large results without running out of memory.

### CODE REVIEW

File: crud/src/main/java/com/redhat/lightblue/assoc/CompositeFindImpl.java [hunk 1]
```diff
@@ -140,7 +144,9 @@ public void explain(OperationContext ctx,
         initialize(ctx,req);
         ExecutionContext executionContext = new ExecutionContext(ctx,null);
         JsonDoc doc=new JsonDoc(executionPlan.explain(executionContext));
-        ctx.addDocument(doc);        
+        ArrayList<DocCtx> l=new ArrayList<>(1);
+        l.add(new DocCtx(doc));
+        ctx.setDocumentStream(new ListDocumentStream<DocCtx>(l));
```
Comment (bserdar) [inline crud/src/main/java/com/redhat/lightblue/assoc/CompositeFindImpl.java]: done

File: crud/src/main/java/com/redhat/lightblue/assoc/ep/StepResultDocumentStream.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.Iterator;
+import java.util.ArrayList;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import com.redhat.lightblue.crud.DocCtx;
+import com.redhat.lightblue.crud.DocumentStream;
+
+/**
+ * Interface between DocumentStream used in CRUDOperationContext, and StepResult
```
Comment (bserdar) [inline crud/src/main/java/com/redhat/lightblue/assoc/ep/StepResultDocumentStream.java]: Not java interface, but an interface between two different streams

File: crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+public interface DocumentStream<T> extends Iterator<T>{
+    /**
+     * Close the document stream
+     */
+    void close();
+
+    /**
+     * Adds a listener that will be called for each document of the stream when next() is called
+     */
+    void tee(Consumer<T> dest);
```
Comment (bserdar) [inline crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java]: tee, as in pipes, and unix tee

File: crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+    /**
+     * returns true if rewind works
+     */
+    default boolean canRewind() { return false; }
+
+    /**
+     * returns a new stream that starts the same resultset from the beginning. Only works if canRewind() is true
+     */
+    default DocumentStream<T> rewind() { throw new UnsupportedOperationException(); }
```
Comment (bserdar) [inline crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java]: Added RewindableDocumentStream

File: crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    public static <S,D> DocumentStream<D> map(final DocumentStream<S> source,final Function<S,D> map) {
```
Comment (bserdar) [inline crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java]: done

File: crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        return new DocumentStreamMapper<S,D>(source,map);
+    }
+
+    static final class DocumentStreamMapper<S,D> implements DocumentStream<D> {
```
Comment (bserdar) [inline crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java]: not final any more. Why separate file?

File: crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java (hunk 3 above)
Comment (dcrissman) [inline crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java]: I still don't think 'tee' is a very good method name.

**Comment by bserdar** [comment]
Fine, renamed. But that's exactly what it does:

https://en.wikipedia.org/wiki/Tee_(command)

File: crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+public interface DocumentStream<T> extends Iterator<T>{
+    /**
+     * Close the document stream
+     */
+    void close();
+
+    /**
+     * Adds a listener that will be called for each document of the stream when next() is called
+     */
+    void forEach(Consumer<T> dest);
```
Comment (dcrissman) [inline crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java]: What about add() or append()?

**Comment by bserdar** [comment]
Like addListener()?

But this isn't exactly a listener. It is a function that will be called for
each doc, as they are pulled from the stream, so forEach tells what it does

**Comment by bserdar** [comment]
Maybe I can change the comment ?

File: crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java (hunk 7 above)
Comment (dcrissman) [inline crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java]: > But this isn't exactly a listener. It is a function that will be called for each doc, as they are pulled from the stream, so forEach tells what it does

Maybe I need to see an example of the method being used? I don't see any calls in this PR, and the only implementations I see just add the Consumer to an array called listeners. I don't even think I see a place that any action is being performed on those arrays.

**Comment by bserdar** [comment]
https://github.com/lightblue-platform/lightblue-core/pull/764/files#diff-1a122660679e1163deac1643961a046fR282

File: crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java (hunk 7 above)
Comment (dcrissman) [inline crud/src/main/java/com/redhat/lightblue/crud/DocumentStream.java]: > I don't even think I see a place that any action is being performed on those arrays.

I take that back, I see it now.

File: crud/src/main/java/com/redhat/lightblue/hooks/HookManager.java [hunk 8]
```diff
@@ -90,25 +92,71 @@ public DocHooks(DocCtx doc, Map<Hook, CRUDHook> hooks) {
...
-    private static final class HookDocs {
-        private final Hook hook;
-        private final CRUDHook crudHook;
-        private final EntityMetadata md;
-        private final List<HookDoc> docs = new ArrayList<>();
+    private class HookAndDocs {
+        final Hook hook;
+        final EntityMetadata md;
+        final CRUDHook resolvedHook;
+        final List<HookDocInfo> docList=new ArrayList<>();;
```
Comment (bserdar) [inline crud/src/main/java/com/redhat/lightblue/hooks/HookManager.java]: fixed

Comment (bserdar) [inline crud/src/main/java/com/redhat/lightblue/hooks/HookManager.java]: Renamed forEach to addListener

**Pull request merged**